File: Remotion/Data/Linq.IntegrationTests.CSharp/LinqSamples101/GroupUserDefinedFunctions.cs
// This file is part of the re-motion Core Framework (www.re-motion.org)
// Copyright (C) 2005-2009 rubicon informationstechnologie gmbh, www.rubicon.eu
// 
// The re-motion Core Framework is free software; you can redistribute it 
// and/or modify it under the terms of the GNU Lesser General Public License 
// as published by the Free Software Foundation; either version 2.1 of the 
// License, or (at your option) any later version.
// 
// re-motion is distributed in the hope that it will be useful, 
// but WITHOUT ANY WARRANTY; without even the implied warranty of 
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
// GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License
// along with re-motion; if not, see http://www.gnu.org/licenses.
// 
using System;
using System.Linq;

namespace Remotion.Data.Linq.IntegrationTests.CSharp.LinqSamples101
{
  internal class GroupUserDefinedFunctions:Executor
  {
    //This sample demonstrates using a scalar user-defined function in a projection.")]
    public void LinqToSqlUserDefined01 ()
    {
      var q = from c in db.Categories
              select new { c.CategoryID, TotalUnitPrice = db.TotalProductUnitPriceByCategory (c.CategoryID) };

      serializer.Serialize (q);
    }

    //This sample demonstrates using a scalar user-defined function in a where clause.")]
    public void LinqToSqlUserDefined02 ()
    {
      var q = from p in db.Products
              where p.UnitPrice == db.MinUnitPriceByCategory (p.CategoryID)
              select p;

      serializer.Serialize (q);
    }

    //This sample demonstrates selecting from a table-valued user-defined function.")]
    public void LinqToSqlUserDefined03 ()
    {
      var q = from p in db.ProductsUnderThisUnitPrice (10.25M)
              where !(p.Discontinued ?? false)
              select p;

      serializer.Serialize (q);
    }

    //This sample demonstrates joining to the results of a table-valued user-defined function.")]
    public void LinqToSqlUserDefined04 ()
    {
      var q = from c in db.Categories
              join p in db.ProductsUnderThisUnitPrice (8.50M) on c.CategoryID equals p.CategoryID into prods
              from p in prods
              select new { c.CategoryID, c.CategoryName, p.ProductName, p.UnitPrice };

      serializer.Serialize (q);
    }
  }
}

File: Remotion/Data/Linq.IntegrationTest.VisualBasic/LinqSamples101/GroupUserDefinedFunctions.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Option Infer On
Option Strict On

Imports System.Collections.Generic
Imports System.Data
Imports System.Data.SqlClient
Imports System.IO
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Reflection
Imports Remotion.Data.Linq.IntegrationTests.TestDomain.Northwind
Imports Remotion.Data.Linq.IntegrationTests.Utilities
Imports System.Data.Linq
Imports System.Text

Namespace LinqSamples101
  Public Class GroupUserDefinedFunctions
    Inherits Executor
    '<Category("User-Defined Functions")> _
    '   <Title("Scalar Function - Select")> _
    '   <Description("This sample demonstrates using a scalar user-defined function in a projection.")> _
    Public Sub LinqToSqlUserDefined01()
      Dim catQuery = From category In db.Categories _
                     Select category.CategoryID, _
                            TotalUnitPrice = db.TotalProductUnitPriceByCategory(category.CategoryID)

      serializer.Serialize(catQuery)
    End Sub

    '<Category("User-Defined Functions")> _
    '<Title("Scalar Function - Where")> _
    '<Description("This sample demonstrates using a scalar user-defined function in a Where clause.")> _
    Public Sub LinqToSqlUserDefined02()

      Dim prodQuery = From prod In db.Products _
                      Where prod.UnitPrice = db.MinUnitPriceByCategory(prod.CategoryID)

      serializer.Serialize(prodQuery)
    End Sub

    '<Category("User-Defined Functions")> _
    '<Title("Table-Valued Function")> _
    '<Description("This sample demonstrates selecting from a table-valued user-defined function.")> _
    Public Sub LinqToSqlUserDefined03()

      Dim prodQuery = From p In db.ProductsUnderThisUnitPrice(10.25D) _
                      Where Not p.Discontinued

      serializer.Serialize(prodQuery)
    End Sub

    '<Category("User-Defined Functions")> _
    '<Title("Table-Valued Function - Join")> _
    '<Description("This sample demonstrates joining to the results of a table-valued user-defined function.")> _
    Public Sub LinqToSqlUserDefined04()

      Dim q = From category In db.Categories _
              Group Join prod In db.ProductsUnderThisUnitPrice(8.5D) _
                    On category.CategoryID Equals prod.CategoryID _
              Into prods = Group _
              From prod2 In prods _
              Select category.CategoryID, category.CategoryName, _
                     prod2.ProductName, prod2.UnitPrice

      serializer.Serialize(q)
    End Sub
  End Class
End Namespace

